### ISSUE TITLE
TP=2: illegal memory access in pg_gather_kernel (parallel/gather.cu) with non-greedy sampling on WSL2 + Blackwell (sm_120)

### ISSUE THREAD

**Issue by Nakanokensetsu** [turboderp-org/exllamav3#245]
## Summary

With tensor-parallel (`TP=2`) on **WSL2 + dual RTX 5070 (Blackwell, sm_120)**, the model loads and runs, but **any request that needs full-vocab logits gathering** — anything other than pure greedy/argmax (i.e. `temperature != 1.0`, `top_p < 1.0`, `top_k > 0`, or repetition/frequency/presence penalty) — **deterministically crashes with an illegal memory access in `pg_gather_kernel` (`exllamav3_ext/parallel/gather.cu`)**. Greedy/argmax sampling *never* crashes.

## Environment

- WSL2 (Ubuntu), 2× RTX 5070 12GB (Blackwell, **sm_120**), driver 596.36, CUDA 12.8
- exllamav3 built from source (~`v0.0.43` / master), `tensor_parallel_backend: nccl`
- No P2P / NVLink (PCIe only); NCCL runs over socket (`lo`)
- Launched with `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True`, `kernel.yama.ptrace_scope=0`, `NCCL_SOCKET_IFNAME=lo`, `NCCL_P2P_DISABLE=1` — this combination is what makes NCCL init succeed on this platform in the first place (legacy CUDA IPC is broken under WSL GPU-PV; `expandable_segments` routes shared handles through `pidfd_getfd`, which needs `ptrace_scope=0`).

## Symptom

Load succeeds (`NCCL warmup finished`), server serves. The first non-greedy request kills a TP worker:

```
GPU assert: an illegal memory access was encountered
  at .../generator/rep_pen.cu:133
```

This reported location is misleading — it's just the first `cudaPeekAtLastError()` reached after an earlier async failure.

## Investigation

1. With `CUDA_LAUNCH_BLOCKING=1`, the real site surfaces in **`parallel/gather.cu`, immediately after `cudaLaunchCooperativeKernel(pg_gather_kernel, …)` in `pg_gather()`**.
2. **Greedy/argmax never crashes** because it takes the `tp_dispatch_lm_head_argmax()` fast path (a small scalar reduce) and skips the full-vocab `pg_gather` entirely. Every sampler that needs full logits goes through `pg_gather`, and all of them hit the same crash — so this is not sampler-specific, it's the gather.
3. The single-case (no-overflow) index math inside the kernel checks out by hand for our (asymmetric) vocab split — no arithmetic OOB found by inspection.
4. **Adding `printf` instrumentation inside `pg_gather_kernel` (no logic change at all) makes the crash location move.** It disappears from `gather.cu` and instead surfaces asynchronously via the NCCL watchdog on the *other* rank; removing the `printf` brings it back to `gather.cu` cleanly and reproducibly. A purely deterministic index bug would not move under instrumentation. This strongly suggests a **timing / memory-ordering-sensitive race** in the producer/consumer handshake — **or** that `cudaLaunchCooperativeKernel` / `grid.sync()` isn't fully honored under WSL2 + sm_120.
5. We couldn't confirm the root cause on this box: a Blackwell-capable `compute-sanitizer` (racecheck) is unavailable (`Device not supported` on sm_120), and `cuda-gdb` was unavailable.

## Minimal repro

`TP=2`, then any of `repetition_penalty != 1.0` / `temperature != 1.0` / `top_p < 1.0` / `top_k > 0` → always crashes in `pg_gather_kernel`. Pure greedy (argmax) → never crashes.

## Workaround (stable for us)

Routing the full-vocab gather (`gather` and `gather_small` in `TPBackendNCCL`) through NCCL point-to-point `dist.send` / `dist.recv`, instead of the custom host-pinned-SHM cooperative-kernel gather, makes TP inference stable for **all** samplers. It also independently fixed **MTP + TP**, which was separately crashing in `pg_gather_small_kernel`. This side-steps the suspect kernel entirely — it doesn't diagnose the root cause. I'm happy to open a PR with this as an opt-in path if it's useful, though it may be more of a platform workaround than something you'd want on by default.

## Question

Is `pg_gather_kernel`'s cooperative-kernel / shared-memory handshake expected to work under WSL2 + Blackwell (sm_120)? Any pointer on where a producer/consumer ordering assumption might break would help a lot. I can run further experiments if you can suggest instrumentation that doesn't perturb timing (the `printf` sensitivity makes that tricky).

**Comment by turboderp** [comment]
Is it possible to try running without `expandable_segments:True`? That would be my main suspect, I think. It's known to be a quirky feature, and I'd especially not trust it implicitly on the paravirtualized driver in WSL2.

Also, does it happen with the native backend? Would like to rule out any bad interaction between NCCL and the custom kernels.

I've fixed a potentially related bug in the kernel (dev branch) but I've not been able to reproduce the symptoms you're seeing. And producer-consumer ordering doesn't sound like a likely explanation for illegal memory access. It might potentially lead to data corruption (consumer reading prematurely from the shared buffer) but leading into the sampling kernels that would at most produce garbage output. This output would then be fed into an array lookup in Python which should result in wrong token output or a CPU-side out-of-bounds error.

I am working on a new single-process TP backend that's more likely to be Windows-friendly, so long-term this problem will probably just go away (maybe replaced by other fun new bugs.)

**Comment by Champollion9012** [comment]
Two things from a native-Linux 2x sm_120 box that might be useful here. Neither is a fix — you
already have the workaround — but both bear on the questions still open in the thread.

**1. compute-sanitizer does work on sm_120, including racecheck.** Your point 5 says a
Blackwell-capable racecheck was unavailable (`Device not supported` on sm_120) and that cuda-gdb was
too, and that is what left @turboderp's ask for "instrumentation that doesn't perturb timing"
unanswered. On CUDA 13.1 with compute-sanitizer 2025.4.1.0, all three tools run on cc 12.0:

```
memcheck    ========= ERROR SUMMARY: 0 errors
racecheck   ========= RACECHECK SUMMARY: 0 hazards displayed (0 errors, 0 warnings)
synccheck   ========= ERROR SUMMARY: 0 errors
```

(rc=0 for each. That is a trivial smoke kernel rather than your repro — all it shows is that the
tool loads and runs on this arch.) So the "not supported" wall looks like a WSL2 or older-toolkit
property rather than an sm_120 one. Given your finding that adding `printf` moves the crash,
racecheck is exactly the non-perturbing instrument the thread was missing, and it should be reachable
on any native-Linux sm_120 box with a recent toolkit.

**2. `--tp_backend` does not change the gather path, so the native-backend control cannot isolate
it.** This is not an inference — `TPBackendNCCL`'s own docstring says it:

> "Operations not currently implemented directly with NCCL, such as broadcast and gather variants,
> delegate to a native shared-memory fallback backend so the rest of the TP code can use one backend
> interface."

In the bodies (master `6ff3a17`), `TPBackendNCCL.gather` is a single statement,
`self.fallback.gather(...)`, and `gather_small` likewise; `self.fallback` is a `TPBackendNative`, and
`TPBackendNative.gather`/`gather_small` are what call `ext.pg_gather`/`ext.pg_gather_small`. So
switching native ↔ nccl changes barriers and all_reduce, not the gather kernel. Consequence for the
experiment: if the crash still follows the backend, the trigger is in the NCCL process group /
all_reduce interaction rather than in `pg_gather` itself — a sharper hypothesis than "is it the
gather".

One more thing worth noting: the NCCL point-to-point gather you used as a workaround is sitting
**commented out** right beneath that delegation in `TPBackendNCCL.gather` (the
`dst_rank = self.active_devices.index(out_device)` block). So it is a path that was already written
once, which seems like an argument for your opt-in patch rather than against it.

I did not try to reproduce the crash itself. `644dd0e819` ("Possibly fixes Windows/WSL2 issues #244
and #245 — aliasing should be a no-op on Linux") is already on master, so a clean run on native Linux
would just confirm what that commit already predicts. A pre-`644dd0e` vs master comparison here would
say something — happy to run that if it would help.
